File: Day #479/solution.py
# Day 479: Generate all permutations via backtracking, picking remaining elements in
# index order so output is lexicographic. Time: O(n! * n), Space: O(n) recursion.


def permutations(nums):
    res = []
    used = [False] * len(nums)
    cur = []

    def backtrack():
        if len(cur) == len(nums):
            res.append(cur[:])
            return
        for i in range(len(nums)):
            if used[i]:
                continue
            used[i] = True
            cur.append(nums[i])
            backtrack()
            cur.pop()
            used[i] = False

    backtrack()
    return res


def main():
    nums = [1, 2, 3]
    res = permutations(nums)
    print("[" + ",".join("[" + ",".join(map(str, p)) + "]" for p in res) + "]")


if __name__ == "__main__":
    main()

File: Day #479/solution.js
// Generate all permutations via backtracking, picking remaining elements in
// index order so output is lexicographic. Time: O(n! * n), Space: O(n) recursion.

function permutations(nums) {
  const res = [];
  const used = new Array(nums.length).fill(false);
  const cur = [];

  function backtrack() {
    if (cur.length === nums.length) {
      res.push(cur.slice());
      return;
    }
    for (let i = 0; i < nums.length; i++) {
      if (used[i]) continue;
      used[i] = true;
      cur.push(nums[i]);
      backtrack();
      cur.pop();
      used[i] = false;
    }
  }

  backtrack();
  return res;
}

function main() {
  const nums = [1, 2, 3];
  const res = permutations(nums);
  console.log("[" + res.map((p) => "[" + p.join(",") + "]").join(",") + "]");
}

main();
